pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+        {
+            LOG (error) << "Checkpoint validation failure - unable to verify envelope";
+            continue;
+        }
+        checkpoint_hashes[checkpoint_t(checkpoint_message.sequence(), checkpoint_message.state_hash())].insert(envelope.sender());
+    }
+
+    //  filter checkpoint_hashes to only those pairs (checkpoint, senders) such that |senders| >= 2f+1
+    std::map<bzn::checkpoint_t , std::set<bzn::uuid_t>> retval;
+    std::copy_if(checkpoint_hashes.begin(), checkpoint_hashes.end(), std::inserter(retval, retval.end()), [&](const auto& h) {return h.second.size() >= 2 * this->max_faulty_nodes() + 1;});
+
+    return retval;
+}
+
+// NOTE: Use validate_and_extract_checkpoint_hashes instead of validate_viewchange_checkpoints
+
+bool
+pbft::is_valid_viewchange_message(const pbft_msg& viewchange_message, const bzn_envelope& original_msg) const
+{
+    if (!this->is_peer(original_msg.sender()))// TODO: Rich - If redundant, keep this check, remove the other
+    {
+        LOG(debug) << "is_valid_viewchange_message - message not from a peer";
+        return false;
+    }
+
+    auto valid_checkpoint_hashes = this->validate_and_extract_checkpoint_hashes(viewchange_message);
+
+    if (valid_checkpoint_hashes.empty())
+    {
+        LOG(error) << "is_valid_viewchange_message - the checkpoint is invalid";
+        return false;
+    }
+    const auto valid_checkpoint = *(valid_checkpoint_hashes.begin());
+
+
+    // TODO: Refactor this block into a new method - Isabel: should extract a validate_prepared_proof method
+    // all the the prepared proofs are valid  (contains a pre prepare and 2f+1 mathcin prepares)
+    for (int i{0}; i < viewchange_message.prepared_proofs_size(); ++i)
+    {
+        const bzn_envelope& pre_prepare_envelope{viewchange_message.prepared_proofs(i).pre_prepare()};
+        if (!this->is_peer(pre_prepare_envelope.sender()) || !this->crypto->verify(pre_prepare_envelope))
+        {
+            LOG(error) << "is_valid_viewchange_message - a pre prepare message has a bad envelope, or the sender is not in the peers list";
+            return false;
+        }
+
+        pbft_msg preprepare_message;
+        if (!preprepare_message.ParseFromString(pre_prepare_envelope.pbft()) || (preprepare_message.sequence() <= valid_checkpoint.first.first))
+        {
+            LOG(error) << "is_valid_viewchange_message - a pre prepare message has an invalid sequence number, or is malformed";
+            return false;
+        }
+
+        std::set<uuid_t> senders;
+        for (int j{0}; j < viewchange_message.prepared_proofs(i).prepare_size(); ++j)
+        {
+            bzn_envelope prepare_envelope{viewchange_message.prepared_proofs(i).prepare(j)};
+            if (/*!prepare_envelope.ParseFromString(viewchange_message.prepared_proofs(i).prepare(j)) ||*/ !this->is_peer(prepare_envelope.sender()) || !this->crypto->verify(prepare_envelope))
```

Comment:
remove commented code or change to whatever it's supposed to be...

Reply:
Done. Removed the comment.